### PR TITLE
Handle non *os.File buffers. (Fix #207)

### LINKED ISSUE #207: pdfcpu: missing write file descriptor
Hi!

This is breaking change https://github.com/pdfcpu/pdfcpu/commit/d8b8be9e0a1ee44fce6cb546ea0c0c129d5fca7f#diff-1f58690efd2b9d45abf9d269399235baR1031-R1034 on 0.3.4 release.  We are passing buffer based io.Writer and we have no intention to write any files.
Based on my quick analysis, it seems that we are forced to pass `*os.File`, otherwise Write will always fail on `setFileSizeOfWrittenFile`.

Please consider allowing to pass in-memory buffers that are not file associated.

As a workaround, we are reverting to 0.3.3 version.

### PR DESCRIPTION
**Pull request by jozuenoon**
Related to https://github.com/pdfcpu/pdfcpu/issues/207.

### CODE REVIEW

**Comment by jozuenoon** [comment]
Code snippet that fails on 0.3.4:
```
package main

import (
	"bytes"
	"io"

	"github.com/mattetti/filebuffer"
	"github.com/pdfcpu/pdfcpu/pkg/api"

	"github.com/jung-kurt/gofpdf"
)

func main() {

	pdfs := []io.ReadSeeker{
		generatePDF("example 1"),
		generatePDF("example 2"),
	}

	newPDF := &bytes.Buffer{}

	if err := api.Merge(pdfs, newPDF, nil); err != nil {
		panic(err)
	}
}

func generatePDF(text string) io.ReadSeeker {
	var buf bytes.Buffer

	pdf := gofpdf.New("P", "mm", "A4", "")
	pdf.AddPage()
	pdf.SetFont("Arial", "", 14)
	pdf.Text(10, 50, text)

	if err := pdf.Output(&buf); err != nil {
		panic(err)
	}

	return filebuffer.New(buf.Bytes())
}
```

File: pkg/pdfcpu/context.go [hunk 1]
```diff
@@ -719,7 +720,7 @@ type WriteContext struct {
 
 	// The PDF-File which gets generated.
 	*bufio.Writer                     // A writer associated with Fp.
-	Fp                  io.Writer     // A file pointer needed for detecting FileSize.
+	Fp                  *os.File      // A file pointer needed for detecting FileSize.
```
Comment (jozuenoon) [inline pkg/pdfcpu/context.go:723]: Seems like Fp should be *os.File as the comment suggests.

File: pkg/pdfcpu/write.go [hunk 2]
```diff
@@ -31,23 +31,20 @@ import (
...
 		log.CLI.Printf("writing to %s\n", fileName)
 
-		file, err = os.Create(fileName)
+		file, err := os.Create(fileName)
 		if err != nil {
 			return errors.Wrapf(err, "can't create %s\n%s", fileName, err)
 		}
 
 		ctx.Write.Writer = bufio.NewWriter(file)
+		ctx.Write.Fp = file
```
Comment (jozuenoon) [inline pkg/pdfcpu/write.go:47]: I think that we can safely override Fp while Writer is nil.

File: pkg/pdfcpu/write.go [hunk 3]
```diff
@@ -127,7 +124,7 @@ func Write(ctx *Context) error {
 		return err
 	}
 
-	err = setFileSizeOfWrittenFile(ctx.Write, file)
+	err = setFileSizeOfWrittenFile(ctx.Write)
```
Comment (jozuenoon) [inline pkg/pdfcpu/write.go:127]: Since created Fp is inside context, we can drop the file.

File: pkg/pdfcpu/write.go [hunk 4]
```diff
@@ -1017,24 +1014,18 @@ func writeXRef(ctx *Context) error {
...
 
-	f1 := f
-
 	// If writing is Writer based then f is nil.
-	if f == nil {
-		var ok bool
-		if f1, ok = w.Fp.(*os.File); !ok {
-			return errors.New("pdfcpu: missing write file descriptor")
-		}
+	if w.Fp == nil {
```
Comment (jozuenoon) [inline pkg/pdfcpu/write.go:1024]: Now if we send io.Writer (non *os.File) we need to exit here.

File: pkg/pdfcpu/write.go [hunk 5]
```diff
@@ -1017,24 +1014,18 @@ func writeXRef(ctx *Context) error {
...
 	if err != nil {
 		return err
 	}
 
-	f1 := f
-
 	// If writing is Writer based then f is nil.
-	if f == nil {
-		var ok bool
-		if f1, ok = w.Fp.(*os.File); !ok {
```
Comment (jozuenoon) [inline pkg/pdfcpu/write.go:1032]: When io.Writer was non *os.File thw w.Fp was nil.

**Comment by lgonzalez-silen** [comment]
I can confirm that this PR fixes Merge for me locally. Thanks for the PR!

**Comment by hhrutter** [comment]
Good to know. 
I am still testing.

File: pkg/pdfcpu/write.go (hunk 2 above)
Comment (hhrutter) [reply pkg/pdfcpu/write.go:47]: We don't really need this, do we?

**Comment by hhrutter** [comment]
Thanks for discovering this! 💚

**Pull request merged**